lib/uiautomator.js

```diff
@@ -1,225 +1,108 @@
-      logger.error("Unable to spawn adb: " + err.message);
-      if (!this.alreadyExited) {
-        this.alreadyExited = true;
-        readyCb(new Error("Unable to start Android Debug Bridge: " +
-          err.message));
-      }
-    }.bind(this));
-    this.proc.stdout.on('data', this.outputStreamHandler.bind(this));
-    this.proc.stderr.on('data', this.errorStreamHandler.bind(this));
-    this.proc.on('exit', this.exitHandler.bind(this));
-  }.bind(this));
-};
-
-UiAutomator.prototype.setExitHandler = function (onExit) {
-  this.onExit = onExit;
-};
-
-UiAutomator.prototype.exitHandler = function (code) {
-  this.cmdCb = null;
-  if (this.socketClient) {
-    this.socketClient.end();
-    this.socketClient.destroy();
-    this.socketClient = null;
+import events from 'events';
+import { SubProcess } from 'teen_process';
+import log from './logger';
+
+class UiAutomator extends events.EventEmitter {
+  constructor (adb) {
+    super();
+    this.adb = adb;
+    this.tempPath = "/data/local/tmp/";
   }
 
-  if (!this.alreadyExited) {
-    this.alreadyExited = true;
-    if (this.shuttingDown) {
-      logger.info("UiAutomator shut down normally");
+  async start (uiAutomatorBinaryPath, className, extraParams, startDetector) {
+    let processIsAlive;
+    if (!startDetector) {
+      startDetector = null;
     }
-    this.onExit(code);
-  }
-};
-
-UiAutomator.prototype.checkForSocketReady = function (output) {
-  if (/Appium Socket Server Ready/.test(output)) {
-    this.socketClient = net.connect(this.systemPort, function () {
-      this.debug("Connected!");
-      this.onSocketReady(null);
-    }.bind(this));
-    this.socketClient.setEncoding('utf8');
-    var oldData = '';
-    this.socketClient.on('data', function (data) {
-      this.debug("Received command result from bootstrap");
-      try {
-        data = JSON.parse(oldData + data);
-        oldData = '';
-      } catch (e) {
-        logger.info("Stream still not complete, waiting");
-        oldData += data;
-        return;
+    try {
+      this.changeState(UiAutomator.STATE_STARTING);
+      log.info("Starting UiAutomator");
+      log.info("Parsing uiautomator jar");
+      // expecting a path like /ads/ads/foo.jar or \asd\asd\foo.jar
+      let jarName = this.parseJarNameFromPath(uiAutomatorBinaryPath);
+      await this.adb.push(uiAutomatorBinaryPath, this.tempPath);
+      // killing any uiautomator existing processes
+      await this.killUiAutomatorOnDevice();
+
+      // for bootstrapJar
+      // jarName = bootstrap.jar
+      // className = io.appium.android.bootstrap.Bootstrap
+      // extraParams = "-e pkg appPackage -e disableAndroidWatchers true/false"
+      // startDetector = (stdout) => {
+      //  if (/Appium Socket Server Ready/.test(stdout)) {
+      //       return true;
+      //  }}
+      let args = ["shell", "uiautomator", "runtest", jarName, "-c", className];
+      if (extraParams) {
+        args.push(extraParams);
```

Comment:
but `args.push(...['foo', 'bar'])` is the same as `args.push('foo',
'bar')`. The array is "spread" into individual arguments to `push` (
https://github.com/lukehoban/es6features#default--rest--spread).

On Fri, Jul 10, 2015 at 1:10 PM, Moiz Virani <notifications@github.com>
wrote:

> In lib/uiautomator.js
> <https://github.com/appium/appium-uiautomator/pull/2#discussion_r34391613>
> :
>
> > +      let jarName = this.parseJarNameFromPath(uiAutomatorBinaryPath);
> > +      await this.adb.push(uiAutomatorBinaryPath, this.tempPath);
> > +      // killing any uiautomator existing processes
> > +      await this.killUiAutomatorOnDevice();
> > +
> > +      // for bootstrapJar
> > +      // jarName = bootstrap.jar
> > +      // className = io.appium.android.bootstrap.Bootstrap
> > +      // extraParams = "-e pkg appPackage -e disableAndroidWatchers true/false"
> > +      // startDetector = (stdout) => {
> > +      //  if (/Appium Socket Server Ready/.test(stdout)) {
> > +      //       return true;
> > +      //  }}
> > +      let args = ["shell", "uiautomator", "runtest", jarName, "-c", className];
> > +      if (extraParams) {
> > +        args.push(extraParams);
>
> args.push(['foo', 'bar']) will be [args..., ['foo', 'bar']] will push
> array as last element in the array
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/appium/appium-uiautomator/pull/2/files#r34391613>.
>